File: src/02.string/04.transform/04.539.go
package transform

import (
	"sort"
	"strconv"
	"strings"
)

func minF(a, b int) int {
	if a > b {
		return b
	}
	return a
}

func FindMinDifference(timePoints []string) int {
	n := len(timePoints)
	maxTime := 24 * 60

	if n > maxTime {
		return 0
	}

	timeMinutes := make([]int, n)
	for i, v := range timePoints {
		item := strings.Split(v, ":")
		hour, _ := strconv.Atoi(item[0])
		minute, _ := strconv.Atoi(item[1])
		timeMinutes[i] = hour*60 + minute
	}

	sort.Ints(timeMinutes)

	min := minF(timeMinutes[0]+maxTime-timeMinutes[n-1], maxTime)
	for i := 0; i < n-1; i++ {
		min = minF(min, timeMinutes[i+1]-timeMinutes[i])
	}

	return min
}

File: src/02.string/04.transform/04.539.ts
export function findMinDifference(timePoints: string[]): number {
    const n = timePoints.length;
    const maxTime = 24 * 60;

    // 抽屉原理, 总计共有 1440 个值, 如果超过 1440 个值, 肯定有两个一样的
    if (n > maxTime) {
        return 0;
    }

    // 转为分钟
    const timeMinutes: number[] = [];
    for (let i = 0; i < n; i++) {
        const item = timePoints[i].split(":");
        timeMinutes[i] = Number(item[0]) * 60 + Number(item[1]);
    }

    timeMinutes.sort((a, b) => a - b);

    let min = Math.min(timeMinutes[0] + maxTime - timeMinutes[n - 1], maxTime);
    for (let i = 0; i < n - 1; i++) {
        min = Math.min(timeMinutes[i + 1] - timeMinutes[i], min);
    }

    return min;
}
